<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Fast single-cell data analysis with non-negative matrix factorization.">
<title>singlet • singlet</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="singlet">
<meta property="og:description" content="Fast single-cell data analysis with non-negative matrix factorization.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">singlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.99</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/Batch_Integration_with_Singlet.html">Batch Integration with Singlet</a>
    <a class="dropdown-item" href="articles/Guided_Clustering_with_NMF.html">Guided Clustering with NMF</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/zdebruine/singlet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="singlet-v0099">singlet v.0.0.99<a class="anchor" aria-label="anchor" href="#singlet-v0099"></a>
</h1></div>
<p><strong>singlet is being actively developed (including documentation) and there are still plenty of known bugs, including in the LNMF function</strong>.</p>
<p>Singlet brings fast Non-negative Matrix Factorization, new integration methods, and more to every Seurat user.</p>
<div class="section level2">
<h2 id="get-started">Get Started!<a class="anchor" aria-label="anchor" href="#get-started"></a>
</h2>
<pre class="{r}"><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"singlet"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zdebruine/singlet" class="external-link">singlet</a></span><span class="op">)</span></span></code></pre>
<p>Planned vignettes: * Guided clustering tutorial with <code>SeuratData::pbmc3k</code> dataset * Batch integration with <code>SeuratData::ifnb</code> dataset (PBMC +/- stimulation) * Multi-modal integration with <code>SeuratData::bmcite</code></p>
</div>
<div class="section level2">
<h2 id="dimensional-reduction-with-nmf">Dimensional Reduction with NMF<a class="anchor" aria-label="anchor" href="#dimensional-reduction-with-nmf"></a>
</h2>
<p>Analyze your single-cell assay with NMF:</p>
<pre class="{r}"><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zdebruine/singlet" class="external-link">singlet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/singlet/man/RunNMF.html" class="external-link">RunNMF</a></span><span class="op">(</span><span class="va">pbmc3k</span>, k <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">20</span>, <span class="fl">2</span><span class="op">)</span>, n_replicates <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pbmc3k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc3k</span>, reduction <span class="op">=</span> <span class="st">"nmf"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">nmf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu"><a href="reference/RankPlot.html">RankPlot</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc3k</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span>, labels <span class="op">=</span> <span class="st">"auto"</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
<div class="figure">
<img src="https://github.com/zdebruine/singlet/blob/main/readme_figures/Picture1.png" alt=""><p class="caption">NMF workflow</p>
</div>
<p>NMF can do almost anything that PCA can do, but also: * imputes missing signal * always has an optimal rank (for variance-stabilized data) * uses all the information in your assay (incl. “non-variable” genes) * is robust across experiments * learns signatures of transcriptional activity * is colinear and non-negative (interpretable), rather than orthogonal and signed (not interpretable)</p>
<p>Singlet internally provides the <strong>fastest implementation of NMF</strong>. Cross-validation can take a few minutes for datasets with a few ten thousand cells, but is extremely scalable and runs excellently on HPC nodes and average laptops alike.</p>
</div>
<div class="section level2">
<h2 id="integration-with-linked-nmf">Integration with Linked NMF<a class="anchor" aria-label="anchor" href="#integration-with-linked-nmf"></a>
</h2>
<p>Learn an integrated model of information across modalities or sample batches and explore <strong>shared and unique</strong> signals in each of your groups.</p>
<pre class="{r}"><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zdebruine/singlet" class="external-link">singlet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/singlet/man/RunNMF.html" class="external-link">RunNMF</a></span><span class="op">(</span><span class="va">ifnb</span>, k <span class="op">=</span> <span class="fl">30</span>, split.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RunLNMF.html">RunLNMF</a></span><span class="op">(</span><span class="va">ifnb</span>, split.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="va">jnmf_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RunLNMF.html">MetadataPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, split.by <span class="op">=</span> <span class="st">"stim"</span>, reduction <span class="op">=</span> <span class="st">"nmf"</span><span class="op">)</span></span>
<span><span class="va">lnmf_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RunLNMF.html">MetadataPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, split.by <span class="op">=</span> <span class="st">"stim"</span>, reduction <span class="op">=</span> <span class="st">"lnmf"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"nmf"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">nmf</span><span class="op">)</span>, reduction.name <span class="op">=</span> <span class="st">"jnmf_all"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"lnmf"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">lnmf</span><span class="op">)</span>, reduction.name <span class="op">=</span> <span class="st">"lnmf_all"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"lnmf"</span>, dims <span class="op">=</span> <span class="fu"><a href="reference/RunLNMF.html">GetSharedFactors</a></span><span class="op">(</span><span class="va">ifnb</span>, split.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span>, reduction.name <span class="op">=</span> <span class="st">"lnmf_shared"</span><span class="op">)</span></span>
<span><span class="va">selected_factors</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">lnmf</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">link_matrix</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"nmf"</span>, dims <span class="op">=</span> <span class="va">selected_factors</span>, reduction.name <span class="op">=</span> <span class="st">"jnmf_shared"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">jnmf_plot</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"joint NMF factor/batch weights"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>, <span class="va">lnmf_plot</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"linked NMF factor/batch weights"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>, <span class="fu"><a href="https://wilkelab.org/cowplot/reference/get_legend.html" class="external-link">get_legend</a></span><span class="op">(</span><span class="va">jnmf_plot</span><span class="op">)</span>, rel_widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.4</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_jnmf_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"jnmf_all"</span>, group.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="va">p_jnmf_shared_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"jnmf_shared"</span>, group.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="va">p_lnmf_all_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"lnmf_all"</span>, group.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span><span class="va">p_lnmf_shared_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"lnmf_shared"</span>, group.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="va">p_jnmf_umap</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"joint NMF"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>, </span>
<span>  <span class="va">p_jnmf_shared_umap</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"joint NMF, selected factors"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>, </span>
<span>  <span class="va">p_lnmf_all_umap</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"linked NMF, all factors"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>, </span>
<span>  <span class="va">p_lnmf_shared_umap</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"linked NMF, shared factors"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>, </span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">2</span></span>
<span> <span class="op">)</span>, <span class="fu"><a href="https://wilkelab.org/cowplot/reference/get_legend.html" class="external-link">get_legend</a></span><span class="op">(</span><span class="va">p_jnmf_umap</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span>, rel_widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>Because unique and shared features are now completely separated, we can run GSEA on these features to better understand whether they are technical “batch effects” or biological differences:</p>
<pre><code><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RunGSEA.html">RunGSEA</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"lnmf"</span>, dims <span class="op">=</span> <span class="fu"><a href="reference/RunLNMF.html">GetUniqueFactors</a></span><span class="op">(</span><span class="va">ifnb</span>, split.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/GSEAHeatmap.html">GSEAHeatmap</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"lnmf"</span><span class="op">)</span></span></code></pre>
<div class="figure">
<img src="https://github.com/zdebruine/singlet/blob/main/readme_figures/Picture2.png" alt=""><p class="caption">Integration with NMF</p>
</div>
<p>Unlike Seurat anchor-based methods, integration with LNMF preserves unique signals in the reduction and thus allows you to understand both the <strong>shared and unique</strong> signals in your different modalities/experiments.</p>
<p>Unlike LIGER integrative NMF, integration with LNMF completely uncouples shared and unique signals from one another and does not assume that both shared and unique signals in any given factor can be mapped to a given sample by the same weight (and thus correspond linearly and additively to one another).</p>
<p>Why not LNMF? If signals are highly disparate, they will not be aligned. Work is ongoing to provide initializations to LNMF that can align even the most disparate signals. However, care should be taken to properly interpret results and avoid “forcing” too much integration when signals are radically different.</p>
</div>
<div class="section level2">
<h2 id="ongoing-work">Ongoing Work<a class="anchor" aria-label="anchor" href="#ongoing-work"></a>
</h2>
<p>Singlet is being actively developed:</p>
<ul>
<li>A new single-cell data class that uses 10x less memory than SCE or Seurat (and much faster)</li>
<li>Full support for Seurat and SingleCellExperiment classes</li>
<li>Out-of-core dimensional reduction with NMF</li>
<li>Regularization and weighting to enable discovery of robust transcriptional signatures with NMF</li>
<li>Spatially-aware dimensional reduction</li>
<li>Graph-regularization for more aligned joint NMF models to initialize LNMF</li>
<li>Integrative transfer learning with linked Non-negative Least Squares projections</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/zdebruine/singlet/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/zdebruine/singlet/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 2)</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing singlet</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Zach DeBruine <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-2234-4827" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Zach DeBruine.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
